stm32: Add STM32H747I-DISCO board

andrewleech opened the pull request:
### Summary

Adds board definition for the STM32H747I-DISCO evaluation board.

The STM32H747I-DISCO features a dual-core STM32H747XIH6 MCU with Cortex-M7 @ 400MHz and Cortex-M4 @ 200MHz, providing a platform for evaluating STM32H7 dual-core capabilities with OpenAMP inter-processor communication.

**Hardware features:**
- 32MB SDRAM (IS42S32800G, 32-bit bus @ 120MHz)
- Dual QSPI flash (2x512Mbit MT25QL512ABB = 128MB total)
- USB High-Speed via ULPI PHY (USB3320C-EZK @ 480Mbps)
- Ethernet 10/100 via RMII
- SD card slot with 8-bit SDMMC interface
- 4 LEDs, 1 user button
- ST-LINK/V3E for programming and UART console

**MicroPython configuration:**
- Single-precision floating point
- External QSPI flash for filesystem (32MB allocated)
- USB CDC console and mass storage
- lwIP networking with mbedTLS
- OpenAMP for dual-core M7-M4 communication
- RTC using LSE (32.768kHz crystal on PC14/PC15)

### Testing

Tested on STM32H747I-DISCO rev D-03 board:

**Verified working:**
- Firmware build and flash via STM32CubeProgrammer
- MicroPython REPL on ST-LINK UART (115200 baud, /dev/ttyACM5)
- USB High-Speed enumeration at 480Mbps via ULPI PHY
- USB CDC console
- USB mass storage (32MB QSPI flash)
- 32MB SDRAM access (~33MB free memory)
- GPIO (LED control on PI12-PI15)
- User button (PC13)
- mpremote connectivity
- I2C (WM8994 audio codec with 16-bit addressing)
- ADC (channels PA6_C, PC2_C, PF10)
- SD card (8-bit SDMMC1 interface, pyb.SDCard API)
- SPI (SPI5)
- UART (UART8)

**Known limitations:**
- Ethernet RMII interface requires hardware modification (pin conflict with MEMS microphone on PC1)
- OpenAMP dual-core communication not tested

**Test suite:**
A test suite with 31 tests covering I2C, ADC, SD card, SPI, UART, Ethernet, DAC, PWM/Timers, LEDs and buttons is available (attached as zip file).

### Trade-offs and Alternatives

**LSE Clock:** This PR includes a fix for LSE initialization on STM32H7/F4/F7 families (commit 2/2). The MicroPython code was missing the call to enable the LSE oscillator before selecting it as the RTC clock source. The board now uses LSE for accurate RTC operation.

**USB Configuration:** Board uses USB High-Speed via external ULPI PHY (USB3320C) with the STM32 HAL USB device stack, providing 480Mbps operation and better performance for USB mass storage and CDC transfers compared to Full-Speed operation.

**SDRAM Allocation:** 32MB SDRAM configured for MicroPython heap via split heap configuration, providing large memory space for applications while maintaining filesystem on external QSPI flash.

andrewleech commented:
Test suite attached below (31 tests covering I2C, ADC, SD card, SPI, UART, Ethernet, DAC, PWM/Timers, LEDs and buttons).

andrewleech commented:
Board image PR: https://github.com/micropython/micropython-media/pull/102

dpgeorge commented on ports/stm32/boards/STM32H747I_DISCO/README.md:
```diff
@@ -0,0 +1,102 @@
+# STM32H747I-DISCO Board
```
This README is too much, none of the other boards have it.

Any critical info (eg the Ethernet hardware mod) could be kept, but would be best in `board.md` so it appears on the download page.

dpgeorge commented on ports/stm32/boards/STM32H747I_DISCO/mpconfigboard.h:
```diff
@@ -0,0 +1,326 @@
+#define MICROPY_HW_FMC_D31          (pin_I10)
+
+#define MICROPY_HW_USB_VID                      0x0483
+#define MICROPY_HW_USB_PID                      0x5740
```
Why a custom VID/PID?

dpgeorge commented on ports/stm32/boards/STM32H747I_DISCO/mpconfigboard.h:
```diff
@@ -0,0 +1,326 @@
+#define MICROPY_HW_USB_PID_CDC_MSC_HID          (MICROPY_HW_USB_PID)
+#define MICROPY_HW_USB_PID_CDC2_MSC_HID         (MICROPY_HW_USB_PID)
+#define MICROPY_HW_USB_PID_CDC3_MSC_HID         (MICROPY_HW_USB_PID)
+#define MICROPY_HW_USB_LANGID_STRING            0x409
```
This is not needed, it's the default.

I think this was copied from the Portenta H7.  I think just remove all this custom USB config.

dpgeorge commented on ports/stm32/boards/STM32H747I_DISCO/mpconfigboard.h:
```diff
@@ -0,0 +1,304 @@
+
+#define MICROPY_HW_BOARD_NAME       "STM32H747I-DISCO"
+#define MICROPY_HW_MCU_NAME         "STM32H747"
+#define MICROPY_HW_FLASH_FS_LABEL   "STM32H747I-DISCO"
```
I think the board should just use the standard flash label and not define a custom one (following most other boards).

dpgeorge commented on ports/stm32/boards/STM32H747I_DISCO/mpconfigboard.h:
```diff
@@ -0,0 +1,304 @@
+#define MICROPY_HW_ENABLE_TIMER     (1)
+#define MICROPY_HW_ENABLE_SDCARD    (1)
+#define MICROPY_HW_ENABLE_MMCARD    (0)
+#define MICROPY_HW_ENTER_BOOTLOADER_VIA_RESET   (0)
```
Did you find this option was necessary?  It's best to go via reset if possible, to reset WDT.

dpgeorge commented on ports/stm32/boards/STM32H747I_DISCO/pins.csv:
```diff
@@ -0,0 +1,310 @@
+USB_ULPI_D7,PB5
+USB_ULPI_STP,PC0
+USB_ULPI_DIR,PI11
+USB_ULPI_NXT,PH4
```
Do the USB pins need to be exposed with names?  All the peripherals below are not exposed to the user (ETH, SDMMC, FMC).

dpgeorge commented on ports/stm32/boards/STM32H747I_DISCO/stm32h747_disco.ld:
```diff
@@ -0,0 +1,72 @@
+_estack = ORIGIN(RAM) + LENGTH(RAM) - _estack_reserve;
+_sstack = _estack - 16K; /* tunable */
+
+/* RAM extents for the garbage collector */
```
As far as I can tell, this RAM is not used for the GC, so this comment is misleading.

dpgeorge commented on ports/stm32/boards/STM32H747I_DISCO/stm32h747_disco.ld:
```diff
@@ -0,0 +1,72 @@
+
+/* Location of filesystem flash storage - not used with external flash */
+_micropy_hw_internal_flash_storage_start = ORIGIN(FLASH) + LENGTH(FLASH);
+_micropy_hw_internal_flash_storage_end = ORIGIN(FLASH) + LENGTH(FLASH);
```
These are wrong, they leave no room for the firmware.

I suggest just removing them, and the `_micropy_hw_internal_flash_storage_ram_cache_XXX` defines as well, and just requiring that the filesystem be in external flash.

dpgeorge commented on ports/stm32/boards/STM32H747I_DISCO/board_init.c:
```diff
@@ -0,0 +1,64 @@
+            break;
+    }
+
+    #if MICROPY_HW_ENABLE_INTERNAL_FLASH_STORAGE == 0
```
As per comment below, could probably remove this `#if` and just require external storage.

pi-anl committed:
stm32/boards/STM32H747I_DISCO: Add STM32H747I-DISCO board definition.

Adds support for STM32H747I-DISCO evaluation board with:
- STM32H747XI dual-core MCU (Cortex-M7 @ 400MHz + Cortex-M4)
- 32MB SDRAM (32-bit bus @ 120MHz)
- 32MB QSPI flash storage (MT25QL512ABB dual)
- USB High-Speed via ULPI PHY (USB3320C-EZK)
- USART1 console via ST-LINK V3 VCP
- 10/100 Ethernet (LAN8742A RMII) - requires hardware modification
- microSD card slot (8-bit SDMMC1)
- 4 LEDs and user button
- OpenAMP support for M7-M4 communication

Software configuration:
- RTC uses LSE (32.768kHz crystal)
- lwIP networking stack with SSL/TLS (mbedTLS)
- exFAT filesystem support

Known limitations:
- Ethernet requires hardware modification (ETH_MDC/SAI4_D1 pin conflict)

Signed-off-by: Andrew Leech <andrew.leech@planetinnovation.com.au>

dpgeorge commented:
Ping @andrewleech , there are comments above to respond to :smile:

dpgeorge commented:
This all looks good now, except I'm still confused about `pins.csv`.

For example, why is it `PA0,PA0` but then `PA1,-PA1`?

> * PC14/PC15 (LSE) and PH0/PH1 (HSE) now hidden

That's not true, there are still the following entries:
```
PC14,-PC14
PC15,-PC15
...
PH0,-PH0
PH1,-PH1
```
The first column is the board name, so users can still access these pins via `Pin.board.PC14` etc.

I suggest keeping `pins.csv` simple and only adding pins that are accessible on the dev board.  Does it have an Arduino connector on the back?  If so, just expose them like this (copied from STM32F769DISC):
```
A0,PA0
A1,PF10
A2,PF9
A3,PF8
A4,PF7
A5,PF6
D0,PC7
D1,PC6
D2,PG6
D3,PB4
D4,PG7
D5,PA8
D6,PH6
D7,PI3
D8,PI2
D9,PA15
D10,PI0
D11,PB15
D12,PB14
D13,PI1
D14,PB9
D15,PB8
```
And then add the LED etc named pins you've already got.  So, don't include any lines like `PA0,PA0`.

andrewleech commented:
Thanks @dpgeorge, sorry for the slow turnaround on this one. I've reworked pins.csv pretty significantly based on your feedback.

The main change is removing the full PA0-PK7 CPU pin enumeration and replacing it with Arduino connector names (A0-A5, D0-D15) following the STM32F769DISC pattern you suggested. The board does have an Arduino header on the back, verified against the schematic and cross-referenced with the Zephyr device tree.

I went back and forth a fair bit on how much to include beyond the user-facing pins. I ended up going with named hidden entries for the internal peripherals (e.g. `-FMC_D0,PD14`) and referencing them via `pyb_pin_*` in mpconfigboard.h, which keeps the pin-to-function mapping centralised in one place. It does mean pins.csv is still pretty long, but there's a lot of bus pins on this board between 32-bit SDRAM, QSPI, ethernet and USB ULPI.

Also dropped the duplicate LED names (was LED1-4 + LED_GREEN/ORANGE/RED/BLUE), now just LED1-4 matching your F769DISC convention.

Happy to simplify further if this is still too much.

dpgeorge reviewed (approved):
Thanks for updating `pins.csv`, this looks good now.
